### PR TITLE
gRPC calls return streams instead of batches.

### PR DESCRIPTION
**Pull request by brendoncarroll**
This changes ListRepo, ListBranch, and ListPipeline to return streams of those objects instead of batches.
Also adds an internal package `clientsdk` with helper functions for reading from these streams.

### CODE REVIEW

File: src/client/pps.go [hunk 1]
```diff
@@ -709,13 +718,15 @@ func (c APIClient) ListPipeline() ([]*pps.PipelineInfo, error) {
 // 1: Return the above and the next most recent version
 // 2: etc.
 //-1: Return all historical versions.
-func (c APIClient) ListPipelineHistory(pipeline string, history int64) ([]*pps.PipelineInfo, error) {
+func (c APIClient) ListPipelineHistory(pipeline string, history int64) (_ []*pps.PipelineInfo, retErr error) {
```
Comment (Tryneus) [inline src/client/pps.go:721]: `retErr` isn't actually used here

File: src/server/pfs/server/driver.go [hunk 2]
```diff
@@ -1819,16 +1814,12 @@ func (d *driver) addBranchProvenance(txnCtx *txncontext.TransactionContext, bran
 func (d *driver) deleteAll(txnCtx *txncontext.TransactionContext) error {
 	// Note: d.listRepo() doesn't return the 'spec' repo, so it doesn't get
```
Comment (PFedak) [inline src/server/pfs/server/driver.go:1815]: Not related to your change, but I forgot to delete this comment

File: src/internal/clientsdk/pfs.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	if err := ForEachBranchInfo(client, func(x *pfs.BranchInfo) error {
+		results = append(results, x)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+func ForEachRepoInfo(client pfs.API_ListRepoClient, cb func(*pfs.RepoInfo) error) (retErr error) {
```
Comment (brycemcanally) [inline src/internal/clientsdk/pfs.go:36]: `retErr` is unused.

File: src/internal/clientsdk/pps.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package clientsdk
+
+import (
+	"io"
+
+	"github.com/pachyderm/pachyderm/v2/src/pps"
+)
+
+func ForEachPipelineInfo(client pps.API_ListPipelineClient, cb func(*pps.PipelineInfo) error) (retErr error) {
```
Comment (brycemcanally) [inline src/internal/clientsdk/pps.go:9]: retErr

File: src/server/auth/server/testing/auth_test.go [hunk 5]
```diff
@@ -1238,8 +1241,9 @@ func TestListRepoNotLoggedInError(t *testing.T) {
 		buildBindings(alice, auth.RepoOwnerRole), getRepoRoleBinding(t, aliceClient, repo))
 
 	// Anon (non-logged-in user) calls ListRepo, and must receive an error
-	_, err := anonClient.PfsAPIClient.ListRepo(anonClient.Ctx(),
+	c, _ := anonClient.PfsAPIClient.ListRepo(anonClient.Ctx(),
```
Comment (brycemcanally) [inline src/server/auth/server/testing/auth_test.go:1244]: No error check?

File: src/server/cmd/pachctl/cmd/cmd.go [hunk 6]
```diff
@@ -509,11 +510,15 @@ This resets the cluster to its initial state.`,
 			for _, ri := range repoInfos {
 				repos = append(repos, red(ri.Repo.Name))
 			}
-			resp, err := client.PpsAPIClient.ListPipeline(client.Ctx(), &pps.ListPipelineRequest{AllowIncomplete: true})
+			c, err := client.PpsAPIClient.ListPipeline(client.Ctx(), &pps.ListPipelineRequest{AllowIncomplete: true})
 			if err != nil {
 				return err
 			}
-			for _, pi := range resp.PipelineInfo {
+			pipelineInfos, err := clientsdk.ListPipelineInfo(c)
```
Comment (brycemcanally) [inline src/server/cmd/pachctl/cmd/cmd.go:517]: This looks like it should use the callback function.

File: src/server/cmd/pachctl/shell/completions.go [hunk 7]
```diff
@@ -105,7 +106,11 @@ func BranchCompletion(flag, text string, maxCompletions int64) ([]prompt.Suggest
 		if err != nil {
 			return nil, CacheNone
 		}
-		for _, bi := range bis.BranchInfo {
+		bis, err := clientsdk.ListBranchInfo(client)
```
Comment (brycemcanally) [inline src/server/cmd/pachctl/shell/completions.go:109]: Same as above.

File: src/server/cmd/pachctl/shell/completions.go [hunk 8]
```diff
@@ -194,12 +199,16 @@ func FilesystemCompletion(_, text string, maxCompletions int64) ([]prompt.Sugges
 // PipelineCompletion completes pipeline parameters of the form <pipeline>
 func PipelineCompletion(_, _ string, maxCompletions int64) ([]prompt.Suggest, CacheFunc) {
 	c := getPachClient()
-	resp, err := c.PpsAPIClient.ListPipeline(c.Ctx(), &pps.ListPipelineRequest{AllowIncomplete: true})
+	client, err := c.PpsAPIClient.ListPipeline(c.Ctx(), &pps.ListPipelineRequest{AllowIncomplete: true})
+	if err != nil {
+		return nil, CacheNone
+	}
+	pipelineInfos, err := clientsdk.ListPipelineInfo(client)
```
Comment (brycemcanally) [inline src/server/cmd/pachctl/shell/completions.go:206]: Same as above.

File: src/server/pachyderm_test.go [hunk 9]
```diff
@@ -8442,17 +8443,23 @@ func TestNoOutputRepoDoesntCrashPPSMaster(t *testing.T) {
...
+		lpClient, err := c.PpsAPIClient.ListPipeline(
+			ctx,
 			&pps.ListPipelineRequest{
 				AllowIncomplete: true,
 			},
 		)
 		if err != nil {
 			return grpcutil.ScrubGRPC(err)
 		}
+		pipelineInfos, err := clientsdk.ListPipelineInfo(lpClient)
```
Comment (brycemcanally) [inline src/server/pachyderm_test.go:8457]: Same as above.

File: src/server/pfs/cmds/cmds.go [hunk 10]
```diff
@@ -772,11 +773,14 @@ Any pachctl command that can take a Commit ID, can take a branch name instead.`,
 				return err
 			}
 			defer c.Close()
-			branchInfos, err := c.PfsAPIClient.ListBranch(c.Ctx(), &pfs.ListBranchRequest{Repo: cmdutil.ParseRepo(args[0])})
+			branchClient, err := c.PfsAPIClient.ListBranch(c.Ctx(), &pfs.ListBranchRequest{Repo: cmdutil.ParseRepo(args[0])})
+			if err != nil {
+				return err
+			}
+			branches, err := clientsdk.ListBranchInfo(branchClient)
```
Comment (brycemcanally) [inline src/server/pfs/cmds/cmds.go:780]: Same as above.

File: src/server/pps/cmds/cmds.go [hunk 11]
```diff
@@ -729,11 +730,14 @@ All jobs created by a pipeline will create commits in the pipeline's output repo
 			if pipeline != "" {
 				request.Pipeline = pachdclient.NewPipeline(pipeline)
 			}
-			response, err := client.PpsAPIClient.ListPipeline(client.Ctx(), request)
+			lpClient, err := client.PpsAPIClient.ListPipeline(client.Ctx(), request)
+			if err != nil {
+				return grpcutil.ScrubGRPC(err)
```
Comment (brycemcanally) [inline src/server/pps/cmds/cmds.go:735]: We probably need to scrub the other call sites as well.

File: src/server/auth/server/testing/auth_test.go (hunk 5 above)
Comment (brendoncarroll) [reply src/server/auth/server/testing/auth_test.go:1244]: I guess the first error should always be nil.

File: src/server/pps/cmds/cmds.go (hunk 11 above)
Comment (brendoncarroll) [reply src/server/pps/cmds/cmds.go:735]: This is just for errors that go to the user right? I'm looking at what this does, and you would never want to call it for any error you wanted to inspect.

**Review by brycemcanally — APPROVED** [review]
lgtm

**Pull request merged**